File: Tree Structures/Heaps.java
//Java
public class HeapNode {
    int value;
    HeapNode left;
    HeapNode right;
    public HeapNode(int value){
        this.value = value;
        this.left = null;
        this.right = null;
    }
    
}
import java.util.Arrays;

public class MinHeap {
    int size;
    int maxSize;
    int[] heap;
    public MinHeap(int max){
        this.maxSize = max;
        heap = new int[max];
        size = 0; 
    }
    // Helper Methods
    public int getLeftChildIndex(int parentIndex){
        return (2 * parentIndex) + 1;
    }
    public int getRightChildIndex(int parentIndex){
        return (2 * parentIndex) + 2;
    }
    public int getParentIndex(int childIndex){
        return (childIndex - 1)/2;
    }

    public boolean hasLeftChild(int index){
        if (getLeftChildIndex(index) > size){
            return false;
        }
        return true;
    }
    public boolean hasRightChild(int index){
        if (getRightChildIndex(index) > size){
            return false;
        }
        return true;
    }
    public boolean hasParent(int index){
        if (index > size || index == 0){
            return false;
        }
        return true;
    }

    public int leftChild(int index){
        return heap[getLeftChildIndex(index)];
    }
    public int rightChild(int index){
        return heap[getRightChildIndex(index)];
    }
    public int parent(int index){
        return heap[getParentIndex(index)];
    }
    
    public void expandArr(){
        if (size == maxSize){
            heap = Arrays.copyOf(heap, maxSize*2);
            maxSize = maxSize * 2;
        }
    }
    public boolean isEmpty(){
        return size == 0;
    }

    public int peek(){
        if (isEmpty()){
            throw new IllegalStateException();
        }
        return heap[0];
    }

    public int poll(){
        if (isEmpty()){
            throw new IllegalStateException();
        }
        int temp = heap[0];
        heap[0] = heap[size-1];
        size--;
        heapifyDown();
        return temp;

    }

    public void add(int value){
        expandArr();
        heap[size] = value;
        size++;
        heapifyUp();
    }

    public void heapifyDown(){
        int i = 0;
        while (hasLeftChild(i)){
            int smallestIndex = getLeftChildIndex(i);
            if (hasRightChild(i) && (rightChild(i) < leftChild(i))){
                smallestIndex = getRightChildIndex(i);
            }
            if (heap[i] < heap[smallestIndex]){
                return;
            }
            else{
                swap(i, smallestIndex);
            }
            i = smallestIndex;
        }
    }
    
    public void heapifyUp(){
        int i = size-1;
        while(hasParent(i)){
            if (parent(i) > heap[i]){
                swap(getParentIndex(i), i);
            }
            else{
                return;
            }
            i = getParentIndex(i);
        }
    }
    
    public void swap(int index1, int index2){
        int temp = heap[index1];
        heap[index1] = heap[index2];
        heap[index2] = temp;
    }

    public void display(){
        System.out.print("[");
        for (int i = 0; i < size; i++){
            System.out.print(heap[i]);
            if (i != size-1){
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }

    public int[] getHeap(){
        int[] rArr = new int[size];
        for (int i = 0; i < size; i++){
            rArr[i] = heap[i];
        }
        return rArr;
    }

    public int height(){
        return 1 + (int) (Math.log(size) / Math.log(2));
    }
}
public class MinHeapTree {
    HeapNode root;
    public MinHeapTree(){
        root = null;
    }

    public HeapNode convertArrToTree(int[] heap, HeapNode current, int i){
        if (i < heap.length){
            HeapNode temp = new HeapNode(heap[i]);
            current = temp;
            
            current.left = convertArrToTree(heap, current.left, 2 * i + 1);
            current.right = convertArrToTree(heap, current.right, 2 * i + 2);
        }
        return current;
    }

    public void printInOrder(HeapNode current){
        if (current == null){
            return;
        }
        printInOrder(current.left);
        System.out.print(current.value + ", ");
        printInOrder(current.right);
    }
    public void printLevelOrder(int height){
        int i;
        System.out.print("[ ");
        for (i = 0; i <= height; i++){
            printLevel(root, i);
        }
        System.out.println("]");
        
    }

    public void printLevel(HeapNode current, int level){
        if (current == null){
            return;
        }
        if (level == 1){
            System.out.print(current.value + " ");
            
        }
        else if (level > 1){
            printLevel(current.left, level-1);
            printLevel(current.right, level-1);
        }

    }

}
import java.util.ArrayList;

public class main{
    public static void main(String[] args){
        MinHeap h = new MinHeap(10);
        h.add(0);
        h.add(1);
        h.add(2);
        h.add(3);
        h.add(4);
        h.add(5);
        h.add(6);
        h.add(7);
        System.out.println(h.poll());
        h.display();
        System.out.println(h.height());
        MinHeapTree mT = new MinHeapTree();
        mT.root = mT.convertArrToTree(h.getHeap(), mT.root, 0);
        mT.printLevelOrder(h.height());

    }
}

File: Tree Structures/Heaps.py
import math
class HeapNode(object):
    def __init__(self, value, left = None, right = None):
        self.value = value
        self.left = left
        self.right = right

class MinHeap(object):
    def __init__(self, max, size = 0):
        self.maxSize = max
        self.size = size
        self.heap = list()
        self.heap = [None] * self.maxSize
    
    def getLeftChildIndex(self, parentIndex):
        return int((2 * parentIndex) + 1)
    
    def getRightChildIndex(self, parentIndex):
        return int((2 * parentIndex) + 1)
    
    def getParentIndex(self, childIndex):
        return int((childIndex - 1)/2)
    
    def hasLeftChild(self, index):
        if (self.getLeftChildIndex(index) > self.size):
            return False
        return True
    
    def hasRightChild(self, index):
        if (self.getRightChildIndex(index) > self.size):
            return False
        return True
    
    def hasParent(self, index):
        if (index > self.size or index == 0):
            return False
        return True
    
    def leftChild(self, index):
        return self.heap[self.getLeftChildIndex(index)]
    
    def rightChild(self, index):
        return self.heap[self.getRightChildIndex(index)]
    
    def parent(self, index):
        return self.heap[self.getParentIndex(index)]

    def expandArr(self):
        if (self.size == self.maxSize):
            temp = self.heap
            self.heap = [None] * (self.maxSize * 2)
            for x in range(0, len(temp)):
                self.heap[x] = temp[x]
            self.maxSize = self.maxSize * 2
        
    def isEmpty(self):
        return self.size == 0
    
    def peek(self):
        if (self.isEmpty()):
            return "Error"
        return self.heap[0]

    def poll(self):
        if (self.isEmpty()):
            return "Error"
        temp = self.heap[0]
        self.heap[0] = self.heap[self.size-1]
        self.size = self.size - 1
        self.heapifyDown()
        return temp
    
    def add(self, value):
        self.expandArr()
        self.heap[self.size] = value
        self.size = self.size + 1
        self.heapifyUp()

    def heapifyDown(self):
        i = 0
        while(self.hasLeftChild(i)):
            smallestIndex = self.getLeftChildIndex(i)
            if (self.hasRightChild(i) and (self.rightChild(i) < self.leftChild(i))):
                smallestIndex = self.getRightChildIndex(i)
            if (self.heap[i] < self.heap[smallestIndex]):
                return
            else:
                self.swap(i, smallestIndex)
            i = smallestIndex

    def heapifyUp(self):
        i = self.size - 1
        while (self.hasParent(i)):
            if (self.parent(i) > self.heap[i]):
                self.swap(self.getParentIndex(i), i)
            else:
                return
            i = self.getParentIndex(i)

    def swap(self, index1, index2):
        temp = self.heap[index1]
        self.heap[index1] = self.heap[index2]
        self.heap[index2] = temp

    def display(self):
        print("[", end="")
        for x in range(0, self.size):
            print(self.heap[x], end="")
            if (x != self.size - 1):
                print(", ", end="")
        print("]")

    def getHeap(self):
        rArr = list()
        for x in range(0, self.size):
            rArr.append(self.heap[x])
        return rArr

    def height(self):
        return 1 + int(math.log(self.size, 2))
class MinHeapTree(object):
    def __init__(self):
        self.root = None
    
    def convertArrToTree(self, heap, current, i):
        if (i < len(heap)):
            temp = HeapNode(heap[i])
            current = temp

            current.left = self.convertArrToTree(heap, current.left, 2 * i + 1)
            current.right = self.convertArrToTree(heap, current.right, 2 * i + 2)

        return current

    def printInOrder(self, current):
        if (current == None):
            return
        self.printInOrder(current.left)
        print(current.value, end=", ")
        self.printInOrder(current.right)
    
    def printLevelOrder(self, height):
        for x in range (0, height+1):
            self.printLevel(self.root, x)

    def printLevel(self, current, level):
        if (current == None):
            return
        if (level == 1):
            print(str(current.value), end= " ")
        elif(level > 1):
            self.printLevel(current.left, level - 1)
            self.printLevel(current.right, level - 1)
            
h = MinHeap(10)
h.add(0)
h.add(1)
h.add(2)
h.add(3)
h.add(4)
h.add(5)
h.add(6)
h.add(7)
h.poll()
h.display()
mT = MinHeapTree()
mT.root = mT.convertArrToTree(h.getHeap(), mT.root, 0)
mT.printLevelOrder(h.height())
